### PR TITLE
Routing analysis wip - support config object input 

### PR DESCRIPTION
**Pull request by adisos**
- add parsing and example with config object

### CODE REVIEW

File: pkg/ibmvpc/parser.go [hunk 1]
```diff
@@ -150,13 +150,132 @@ func VPCConfigsFromResources(rc *datamodel.ResourcesContainerModel, vpcID, resou
...
 	if debug {
 		printVPCConfigs(res)
 	}
 
 	return res, nil
 }
 
+// getRoutingTables parses routing tables from rc and adds their generated objects to
+// the relevant vpc configs within res
+func getRoutingTables(rc *datamodel.ResourcesContainerModel,
```
Comment (zivnevo) [inline pkg/ibmvpc/parser.go:167]: No `skipByVPC` function?

File: pkg/ibmvpc/parser.go [hunk 2]
```diff
@@ -150,13 +150,132 @@ func VPCConfigsFromResources(rc *datamodel.ResourcesContainerModel, vpcID, resou
...
+		res.SetConfig(vpcUID, vpcConfig)
+	}
+	return nil
+}
+
+func getIngressRoutingTable(rt *datamodel.RoutingTable,
+	routes []*route,
+	vpcConfig *vpcmodel.VPCConfig) (vpcmodel.VPCResourceIntf, error) {
+	if !*rt.RouteTransitGatewayIngress {
+		// todo: skip instead?
```
Comment (zivnevo) [inline pkg/ibmvpc/parser.go:206]: yes, better skip

File: pkg/ibmvpc/parser.go [hunk 3]
```diff
@@ -150,13 +150,132 @@ func VPCConfigsFromResources(rc *datamodel.ResourcesContainerModel, vpcID, resou
...
+		return nil, fmt.Errorf("currently not supporting ingress routing table with source other than tgw ")
+	}
+	res := newIngressRoutingTableFromRoutes(routes, vpcConfig)
+	res.ResourceName = *rt.Name
+	res.ResourceUID = *rt.ID
+	res.ResourceType = ResourceTypeRoutingTable
+	res.VPCRef = vpcConfig.VPC
```
Comment (zivnevo) [inline pkg/ibmvpc/parser.go:213]: consider setting some of these fields in `newIngressRoutingTableFromRoutes()`

File: pkg/ibmvpc/parser.go [hunk 4]
```diff
@@ -150,13 +150,132 @@ func VPCConfigsFromResources(rc *datamodel.ResourcesContainerModel, vpcID, resou
...
+		} else {
+			return nil, fmt.Errorf("could not find subnet %s associated with routing table %s", *s.Name, *rt.Name)
+		}
+	}
+
+	res := newEgressRoutingTableFromRoutes(routes, subnets, vpcConfig)
+	res.ResourceName = *rt.Name
+	res.ResourceUID = *rt.ID
+	res.ResourceType = ResourceTypeRoutingTable
+	res.VPCRef = vpcConfig.VPC
```
Comment (zivnevo) [inline pkg/ibmvpc/parser.go:235]: consider setting some of these fields in `newEgressRoutingTableFromRoutes()`

**Review by zivnevo — CHANGES_REQUESTED** [review]
Overall looks good. Please see some small comments

File: pkg/ibmvpc/routing_tables_analysis.go [hunk 5]
```diff
@@ -484,39 +505,52 @@ func (irt *ingressRoutingTable) advertiseRoutes(vpcConfig *vpcmodel.VPCConfig) {
...
 					vpcB = vpc
 					tgwAB = tgw
+					logging.Debugf("found tgwAB: %s,  vpcB: %s ", tgwAB.Name(), vpcB.Name())
 					break
 				}
 			}
 		}
 		if vpcB == nil {
-			break // nothing to propagate if could not find the relevant vpc and its tgw
+			logging.Debugf(" could not find the relevant vpc and its tgw -- break")
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/routing_tables_analysis.go]: done

File: pkg/ibmvpc/routing_tables_analysis.go [hunk 6]
```diff
@@ -484,39 +505,52 @@ func (irt *ingressRoutingTable) advertiseRoutes(vpcConfig *vpcmodel.VPCConfig) {
...
 		}
 		// find the vpc (with tgw) to which this cidr Y belongs
 		var vpcB *VPC
 		var tgwAB *TransitGateway
 		for _, tgw := range tgws {
 			for _, vpc := range tgw.vpcs {
-				if routeCidr.ContainedIn(vpc.addressPrefixesIPBlock) && vpc.UID() != irt.vpc.UID() {
+				logging.Debugf("check tgw %s with vpc %s, AP %s", tgw.Name(), vpc.Name(), vpc.addressPrefixesIPBlock.ToCidrListString())
+				// TODO: shouldn't be containment rather than intersection?? (works with intersection on hub-n-spoke config object)
+				if vpc.UID() != irt.vpc.UID() && !routeCidr.Intersect(vpc.addressPrefixesIPBlock).IsEmpty() {
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/routing_tables_analysis.go:527]: opened a separate issue #639 (maybe should be added in models repo)

File: pkg/ibmvpc/vpc.go [hunk 7]
```diff
@@ -837,6 +837,15 @@ func (tgw *TransitGateway) ExternalIP() string {
 }
 
 func (tgw *TransitGateway) AllowedConnectivity(src, dst vpcmodel.VPCResourceIntf) (*connection.Set, error) {
+	if _, ok := src.(*vpcmodel.ExternalNetwork); ok {
+		return connection.None(), nil
+	}
+	if _, ok := dst.(*vpcmodel.ExternalNetwork); ok {
+		return connection.None(), nil
+	}
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/vpc.go]: right, changed

File: pkg/common/pair.go [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package common
+
+type Pair[T any] struct {
+	First  T
+	Second T
+}
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/common/pair.go]: right, changed

File: pkg/ibmvpc/parser.go [hunk 9]
```diff
@@ -150,13 +150,132 @@ func VPCConfigsFromResources(rc *datamodel.ResourcesContainerModel, vpcID, resou
...
+		rObj, err := newRoute(*r.Name, *r.Destination, *nextHop.Address, action, int(*r.Priority), *r.Advertise)
+		if err != nil {
+			return nil, err
+		}
+		res = append(res, rObj)
+	}
+	return res, nil
+}
+
+func parseAction(action string) (action, error) {
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/parser.go]: right, changed

File: pkg/ibmvpc/parser.go [hunk 10]
```diff
@@ -150,13 +150,132 @@ func VPCConfigsFromResources(rc *datamodel.ResourcesContainerModel, vpcID, resou
...
+// the relevant vpc configs within res
+func getRoutingTables(rc *datamodel.ResourcesContainerModel,
+	res *vpcmodel.MultipleVPCConfigs) error {
+	for _, rt := range rc.RoutingTableList {
+		routes, err := getRoutes(rt)
+		if err != nil {
+			return err
+		}
+		if rt.VPC == nil || rt.VPC.CRN == nil {
+			logging.Debugf("unknown vpc for rt %s, skipping...", *rt.Name)
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/parser.go]: right, changed

File: pkg/ibmvpc/parser.go [hunk 11]
```diff
@@ -150,13 +150,132 @@ func VPCConfigsFromResources(rc *datamodel.ResourcesContainerModel, vpcID, resou
...
+			continue
+		}
+		vpcUID := *rt.VPC.CRN
+		vpcConfig := res.Config(vpcUID)
+		if vpcConfig == nil {
+			logging.Debugf("skipping rt %s, could not find vpc with uid %s", *rt.Name, vpcUID)
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/parser.go]: right, changed

File: pkg/ibmvpc/parser.go (hunk 1 above)
Comment (adisos) [reply pkg/ibmvpc/parser.go:167]: added

File: pkg/ibmvpc/parser.go (hunk 2 above)
Comment (adisos) [reply pkg/ibmvpc/parser.go]: done

File: pkg/ibmvpc/parser.go (hunk 3 above)
Comment (adisos) [reply pkg/ibmvpc/parser.go]: done

File: pkg/ibmvpc/parser.go (hunk 4 above)
Comment (adisos) [reply pkg/ibmvpc/parser.go]: done

File: pkg/vpcmodel/multi_vpc_config.go [hunk 12]
```diff
@@ -85,3 +85,38 @@ func (c *MultipleVPCConfigs) GetVPC(uid string) VPCResourceIntf {
...
 }
+
+func (c *MultipleVPCConfigs) GetSubnetByUID(uid string) VPCResourceIntf {
+	for _, vpcConfig := range c.configs {
+		if obj, ok := vpcConfig.UIDToResource[uid]; ok {
+			return obj
+		}
+	}
+	return nil
+}
```
Comment (zivnevo) [inline pkg/vpcmodel/multi_vpc_config.go:96]: Is this function needed?

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: pkg/ibmvpc/parser.go [hunk 13]
```diff
@@ -150,13 +150,140 @@ func VPCConfigsFromResources(rc *datamodel.ResourcesContainerModel, vpcID, resou
...
+		if rt.VPC == nil || rt.VPC.CRN == nil {
+			logging.Warnf("unknown vpc for rt %s, skipping...", *rt.Name)
+			continue
+		}
+		vpcUID := *rt.VPC.CRN
+		vpcConfig := res.Config(vpcUID)
+		if vpcConfig == nil {
+			logging.Warnf("skipping rt %s, could not find vpc with uid %s", *rt.Name, vpcUID)
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/parser.go:180]: done

File: pkg/vpcmodel/multi_vpc_config.go (hunk 12 above)
Comment (adisos) [reply pkg/vpcmodel/multi_vpc_config.go]: removed

**Review by zivnevo — APPROVED** [review]
Just remove the "wip" from the commit message 😄

**Pull request merged**